### PR TITLE
iodine service: add client mode implimentation

### PR DESCRIPTION
**Pull request by Pleune**
##### Description

I noticed that the `services.iodined.client` does nothing other than throw an assert under invalid conditions. This is my attempt at adding functionality to the above. However, this has ended up leading to much more than just enabling the option. Old server configurations will continue to work under the name `services.iodined` by rename. However the new configuration lives in `services.iodine`.
###### Server config

A single server can be configured with `services.iodine.server`. It uses the same configuration as the old `services.iodined`.
###### Client config

Any number of clients can now be configured with `services.iodine.clients.*`. They use a very similar configuration as the server. A service called iodine-_name_.service is started for every configuration in place of the astrix above.
##### Tests
- [x] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux

I have tested both the client and server functionality on my computers running NixOS

### CODE REVIEW

File: nixos/modules/services/networking/iodined.nix [hunk 1]
```diff
@@ -62,13 +62,22 @@ in
...
     boot.kernelModules = [ "tun" ];
 
-    systemd.services.iodined = {
+    systemd.services.iodine = mkIf cfg.client {
+      description = "iodine, ip over dns daemon - client";
+      wantedBy = [ "ip-up.target" ];
+      serviceConfig = {
+        RestartSec = "30s";
+        Restart = "always";
+        ExecStart = "${pkgs.iodine}/sbin/iodine -f -u ${iodinedUser} ${cfg.extraConfig} ${cfg.ip} ${cfg.domain}";
```
Comment (joachifm) [inline nixos/modules/services/networking/iodined.nix]: You ought to use `${pkgs.iodine}/bin/iodine`, `/sbin` is legacy.

**Comment by joachifm** [comment]
Superficially, this looks okay to me, but I know nothing about the service nor do I intend to use it.

On a related note: I find the interface somewhat ugly; changing from server to client with a single switch is odd to me, I'd prefer to have `services.iodined-client` and `services.iodined-server` as separate namespaces. Thoughts? (I think we can still merge this in the interim, though, if it works for you).

**Comment by Pleune** [comment]
@joachifm Should I be worried about preserving backwards compatibility then? If not, then my idea would be to create `services.iodined.servers` and `services.iodined.clients`, both as lists. This is similar to how OpenVPN does the same with `services.openvpn.servers`. This way multiple instances can be run with their own configurations, all still configured through `services.iodined`.

**Comment by joachifm** [comment]
There is a mechanism for marking options as deprecated and providing aliases. If that is insufficiently expressive, errors can be thrown instead. Either way, there's no reason to let backwards compatibility cripple improvements more than necessary, IMO. I think this is something users of this module have to discuss, these are just general remarks.

**Comment by Pleune** [comment]
@joachifm 

This is my solution:

The old configuration is 100% backwards compatible. If the old broken client option was set to false, then a warning is thrown. This is because `services.iodined` is renamed to `services.iodine.server` and all the options stay the same (except for the removed `services.iodined.client` boolean).

As for the new configuration, a single server may be enabled and specified with `services.iodine.server` and any number of client services can be spawned through `servies.iodine.clients`.

**Comment by joachifm** [comment]
Okay, cool. Let's leave this opened for a while, though, to let potentially interested parties provide some feedback. Ping me if it ends up lingering for too long; if this works and makes sense to you as a consumer of the service module, then I'm fine with merging it.

**Comment by joachifm** [comment]
Okay, no objections thus far; I think this has been waiting long enough.

**Pull request merged**